src/theory/subs_minimize.cpp

```diff
@@ -0,0 +1,331 @@
+        std::vector<Node> vchildren;
+        if (cur.getMetaKind() == kind::metakind::PARAMETERIZED)
+        {
+          if (cur.getKind() == APPLY_UF)
+          {
+            children.push_back(cur.getOperator());
+          }
+          else
+          {
+            vchildren.push_back(cur.getOperator());
+          }
+        }
+        for (const Node& cn : cur)
+        {
+          children.push_back(cn);
+        }
+        for (const Node& cn : children)
+        {
+          it = value.find(cn);
+          Assert(it != value.end());
+          Assert(!it->second.isNull());
+          vchildren.push_back(it->second);
+        }
+        ret = nm->mkNode(cur.getKind(), vchildren);
+        ret = Rewriter::rewrite(ret);
+      }
+      value[cur] = ret;
+    }
+  } while (!visit.empty());
+  Assert(value.find(n) != value.end());
+  Assert(!value.find(n)->second.isNull());
+
+  Trace("subs-min") << "... got " << value[n] << std::endl;
+  if (value[n] != target)
+  {
+    Trace("subs-min") << "... not equal to target " << target << std::endl;
+    return false;
+  }
+
+  Trace("subs-min") << "--- Compute relevant variables..." << std::endl;
+  std::unordered_set<Node, NodeHashFunction> rlvFv;
+  // only variables that occur in assertions are relevant
+  std::map<Node, unsigned> iteBranch;
+  std::map<Node, std::vector<unsigned> > justifyArgs;
+
+  visit.push_back(n);
+  std::unordered_set<TNode, TNodeHashFunction> visited;
+  std::unordered_set<TNode, TNodeHashFunction>::iterator itv;
+  do
+  {
+    cur = visit.back();
+    visit.pop_back();
+    itv = visited.find(cur);
+    if (itv == visited.end())
+    {
+      visited.insert(cur);
+      it = value.find(cur);
+      if (it->second == cur)
+      {
+        // if its value is the same as current, there is nothing to do
+      }
+      else if (cur.isVar())
+      {
+        // must include
+        rlvFv.insert(cur);
+      }
+      else if (cur.getKind() == ITE)
+      {
+        // only recurse on relevant branch
+        Node bval = value[cur[0]];
+        Assert(!bval.isNull() && bval.isConst());
+        unsigned cindex = bval.getConst<bool>() ? 1 : 2;
+        visit.push_back(cur[0]);
+        visit.push_back(cur[cindex]);
+      }
+      else if (cur.getNumChildren() > 0)
+      {
+        Kind ck = cur.getKind();
+        bool alreadyJustified = false;
+
+        // if the operator is an apply uf, check its value
+        if (cur.getKind() == APPLY_UF)
+        {
+          Node op = cur.getOperator();
+          it = value.find(op);
+          Assert(it != value.end());
+          Node vop = it->second;
+          if (vop.getKind() == LAMBDA)
+          {
+            visit.push_back(op);
+            Node curr = vop[0];
+            for (unsigned i = 0, size = cur.getNumChildren(); i < size; i++)
+            {
+              it = value.find(cur[i]);
+              Assert(it != value.end());
+              Node scurr = curr.substitute(cur[i], it->second);
```

Comment:
Shouldn't this be `Node scurr = vop[1].substitute(curr[i], it->second);`?

Reply:
Yes, I'll fix.